### ISSUE TITLE
Doesn't work with LDAP users.

### ISSUE THREAD

**Issue by Zappatron** [owncloud/impersonate#5]
Impersonate doesn't work here (8.1.3) with LDAP users, using the guid or LDAP user name.

This would be a very useful feature!

Cheers

**Comment by Zappatron** [comment]
I changed the impersonate function like so:

        public function impersonate($userid) {
                $users = $this->userManager->searchDisplayName($userid, 1, 0);
                if (count($users) > 0) {
                        $user = array_shift($users);
                        if (strcasecmp($user->getDisplayName(),$userid) === 0) {
                                $this->userSession->setUser($user);
                        }
                }
                return new JSONResponse();
        }

And it seems to work fine, using Display Name instead of User ID

**Comment by karlitschek** [comment]
@butonic What do you think?

**Comment by Zappatron** [comment]
Obviously this will cause problems with duplicate display names but in our environment this isn't a problem (currently).

I don't know enough about the ownCloud api to know why the user id search does not return the user (or the userid comparison does not match, I didn't investigate which was occuring).

**Comment by scolebrook** [comment]
Here's my version. Tried to do a branch and PR but discovered I'm not a contributor on this repo.

```php
	public function impersonate($userid) {
		if ($this->userManager->userExists($userid)) {
			//$userid is an internal user id
			$newuser = $this->userManager->get($userid);
		} else {
			//$userid is any other search string
			$users = $this->userManager->search($userid, 1, 0);
			if (count($users) > 0 ) {
				$newuser = array_shift($users);
			}
		}
		
		if (!empty($newuser)) {
			//we found something
			$this->userSession->setUser($newuser);
		}
		
		return new JSONResponse();
	}
```

This allows searching by internal uid which might be a username for an internal user, objectguid for an ad user, etc. It also searches for any attribute that the user manager can search. In the case of ldap users, that's all the search attributes that are configured on the admin page. Want to search for displayName? Add it to the search attributes list. Want to search for email address, userPrincipalName, samAccountName, etc? Add them to the list. Anything that can be used to search for users for sharing will work here. The difference is that this will go with the first match rather than give you a choice. So be specific.

Only problem is that it can't search for a display name containing a space. Neither can searching for sharing. I think that's a limitation in user_ldap. If that gets resolved, then this would work with spaces too automajically.

**Comment by butonic** [comment]
@scolebrook fork this repo, check in the changes to your repo, then create a pr from that. Thx!

**Comment by scolebrook** [comment]
@butonic https://github.com/scolebrook/impersonate/pull/1
Sorry it took a while to get back to this.

**Comment by butonic** [comment]
@scolebrook cool ... can you squash the commits and create the pr against this repo? then I can directly merge your changes here!

**Comment by butonic** [comment]
@scolebrook hmmm I just merged code that allows impersonating a user from the user list. that allows you to filter und search for an ldap user as well ... that should allow you to impersonate ldap isers as well. can you verify with master?

**Comment by butonic** [comment]
@Zappatron doas master fix your issue by allowing you to impersonate users from the user list?

**Comment by Zappatron** [comment]
Unfortunately not. Clicking impersonate in the user list takes me to my own files (which was the behaviour behore). According to the log entries below it looks like it's doing an ldap search for displayname but providing the uid:

    Debug	user_ldap	getUsers: 0 Users found	2016-01-07T14:13:58+00:00
    Debug	user_ldap	Ready for a paged search	2016-01-07T14:13:58+00:00
    Debug	user_ldap	initializing paged search for Filter (&(&(|(objectclass=user))(|(|(memberof=CN=ownCloud Users,OU=Resource,OU=Domain Global,OU=Security Groups,DC=XXXXXXXXXX,DC=co,DC=uk)(primaryGroupID=8439))))(displayname=5E8DB41A-A294-4533-B0BC-761B4BC67495*)) base Array ( [0] => dc=XXXXXXXXXX,dc=co,dc=uk ) attr Array ( [0] => displayname [1] => dn ) limit 1 offset 0	2016-01-07T14:13:58+00:00
    Debug	user_ldap	getUsers: Options: search 5E8DB41A-A294-4533-B0BC-761B4BC67495 limit 1 offset 0 Filter: (&(&(|(objectclass=user))(|(|(memberof=CN=ownCloud Users,OU=Resource,OU=Domain Global,OU=Security Groups,DC=XXXXXXXXXX,DC=co,DC=uk)(primaryGroupID=8439))))(displayname=5E8DB41A-A294-4533-B0BC-761B4BC67495*))

**Comment by scolebrook** [comment]
@butonic I see the same as @Zappatron. I think this is a much better UI though. Will the head icon display for someone who has admin rights on a group as opposed to a full admin? That could be an issue.

**Comment by butonic** [comment]
@scolebrook it will currently only be visible to the admin, not sub admins. allowing sub admins would require checking the user to impersonate belongs to the subadmin and I currently don't want to grant imporsonation power to subadmins. PR welcome but it should be configurable by the admin.

**Comment by thomaseitler** [comment]
@butonic Is there any update on this issue? I'm experiencing the same behaviour as the previous posters. It would be really nice to get this to work with LDAP Users as well.
I'm using the stable8.2 branch.

**Comment by butonic** [comment]
@Zappatron @scolebrook @thomaseitler looking at https://github.com/owncloud/impersonate/blob/master/controller/settingscontroller.php#L39

can you try with this implementation instead:
```php
	public function impersonate($userid) {
		if ($this->userManager->userExists($userid)) {
			$this->userSession->setUser($userid);
			return new JSONResponse();
		}
		return new NotFoundResponse();
	}
```

since we are using the usen managemen to search for a user we no longer need to do it here.

**Comment by thomaseitler** [comment]
@butonic I actually tried @scolebrook 's function posted above and got it working. I will try your solution tomorrow and report back.

**Comment by scolebrook** [comment]
@butonic $userid is not a \OCP\IUser object which is required by the session manager but is the string that the user has supplied. Before you can set the session to a given user, you first need to get the user object for the user the admin is looking for. I'm working with 8.1.5. Perhaps this is different in 8.2.x or master. Which version of OC are you working with?

**Comment by butonic** [comment]
fixed by 9af19c4c8860077d4fd7b100680882b994674c9e
